### ISSUE TITLE
Failed to inference grayscale image with InferenceEngine

### ISSUE THREAD

**Issue by michaelnguyen11** [opencv/openvino_training_extensions#143]
Hi Supporters,

I've trained a 1 channel grayscale face detection, followed the instruction with SSD 300x300 mobilenet v2
https://github.com/opencv/openvino_training_extensions/tree/develop/pytorch_toolkit/face_recognition

The training steps work well, but I got a problem when inference the model:
```
terminate called after throwing an instance of 'InferenceEngine::details::InferenceEngineException'
  what():  Failed to infer shapes for Convolution layer (332) with error: New shapes [1,1,300,300] make Kernels(3x3), Channels(1), Output depth(24), Groups(1) not matching weights size: 216 vs 648
```
Coredump info :
```
#6  0x00007fc834901990 in InferenceEngine::details::extract_exception(InferenceEngine::StatusCode, char*) () at /usr/local/lib/libopencv_dnn.so.4.1
#7  0x00007fc8349faf44 in cv::dnn::InfEngineBackendLayer::getMemoryShapes(std::vector<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > > const&, int, std::vector<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > >&, std::vector<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > >&) const () at /usr/local/lib/libopencv_dnn.so.4.1
#8  0x00007fc83490bad9 in cv::dnn::dnn4_v20190621::Net::Impl::getLayerShapesRecursively(int, std::map<int, cv::dnn::dnn4_v20190621::(anonymous namespace)::LayerShapes, std::less<int>, std::allocator<std::pair<int const, cv::dnn::dnn4_v20190621::(anonymous namespace)::LayerShapes> > >&) () at /usr/local/lib/libopencv_dnn.so.4.1
#9  0x00007fc8349200e0 in cv::dnn::dnn4_v20190621::Net::Impl::allocateLayers(std::vector<cv::dnn::dnn4_v20190621::LayerPin, std::allocator<cv::dnn::dnn4_v20190621::LayerPin> > const&) () at /usr/local/lib/libopencv_dnn.so.4.1
#10 0x00007fc8349207d1 in cv::dnn::dnn4_v20190621::Net::Impl::setUpNet(std::vector<cv::dnn::dnn4_v20190621::LayerPin, std::allocator<cv::dnn::dnn4_v20190621::LayerPin> > const&) () at /usr/local/lib/libopencv_dnn.so.4.1
#11 0x00007fc834920ff9 in cv::dnn::dnn4_v20190621::Net::forward(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
    at /usr/local/lib/libopencv_dnn.so.4.1
#12 0x000055c9252434a2
        detection = 
              {flags = 1124024322, dims = 2, rows = 0, cols = 0, data = 0x0, datastart = 0x0, dataend = 0x0, datalimit = 0x0, allocator = 0x0, u = 0x0, size = {p = 0x7fff0613b9e8}, step = {p = 0x7fff0613ba30, buf = {320, 2}}}
        gray = 
              {flags = 1124024320, dims = 2, rows = 120, cols = 160, data = 0x55c926d8ef80 ""..., datastart = 0x55c926d8ef80 ""..., dataend = 0x55c926d93a80 "\256\256\256\255\256\256\255\255\261", datalimit = 0x55c926d93a80 "\256\256\256\255\256\256\255\255\261", allocator = 0x0, u = 0x55c926d3e300, size = {p = 0x7fff0613b8c8}, step = {p = 0x7fff0613b910, buf = {160, 1}}}
        blob = 
              {flags = 1124024325, dims = 4, rows = -1, cols = -1, data = 0x55c92ec48680 "", datastart = 0x55c92ec48680 "", dataend = 0x55c92eca04c0 "\217\217\217\217\217\217\217\217A\023\003", datalimit = 0x55c92eca04c0 "\217\217\217\217\217\217\217\217A\023\003", allocator = 0x0, u = 0x55c926e0a300, size = {p = 0x55c926dee064}, step = {p = 0x55c926dee040, buf = {0, 0}}}
```

The code used to Inference:
```
blob = cv::dnn::blobFromImage(gray, 1.0, Size(300, 300), true, true);
_net.setInput(blob);
cv::Mat detection = _net.forward();
```
Look like something wrong with blob variable : dims = 4, rows = -1, cols = -1
I changed the flag swapRB=false, but still got same issue.

Did you face with the same issue ?
Do you know how to overcome this issue ?

I'm very appreciated your help, thank you in advance !

My Environment:
OS Platform and Distribution : Linux Ubuntu 18.04.3
OpenVino version : openvino_2019.2.242

Regards,
Michael

**Comment by AlexanderDokuchaev** [comment]
Hi @michaelnguyen11 

`rows = -1, cols = -1` is ok for cv::Mat when dimensions more than 3.
`swapRB` is not used for 1 channel images.

May be it's incorrect size of input blob. 
Could you provide your IR?

**Comment by michaelnguyen11** [comment]
Hi @AlexanderDokuchaev ,

Thank for your response.

Please help me check my model and checkpoint :

[IR_and_checkpoint_143.zip](https://github.com/opencv/openvino_training_extensions/files/3577324/IR_and_checkpoint_143.zip)

Regards,
Michael

**Comment by AlexanderDokuchaev** [comment]
@michaelnguyen11  But your model expects input image with 3 channels.
Are you sure that you train model for grayscale image?

**Comment by michaelnguyen11** [comment]
Hi @AlexanderDokuchaev ,

Thank you for your response.

The config that I used  is same as this config : mobilenetv2_tiny_ssd300_wider_face.py, and just change this param:
img_norm_cfg = dict(mean=[0, 0, 0], std=[255, 255, 255], to_rgb=True) to
img_norm_cfg = dict(mean=[0, 0, 0], std=[255, 255, 255], to_rgb=False)

And dataset is all grayscale images. I'm too naive when believe that the output will be for grayscale images.

Could you give me advice how to change the config/code to train grayscale model correctly ?

I'm very appreciated your help, thank you in advance !

Regards,
Michael

**Comment by AlexanderDokuchaev** [comment]
`to_rgb` - used to convert BGR to RGB. [some info why is need](https://www.learnopencv.com/why-does-opencv-use-bgr-color-format/)

> Could you give me advice how to change the config/code to train grayscale model correctly ?

Yes, i can but need some time.
One question, why do you want to use grayscale image?

**Comment by michaelnguyen11** [comment]
Hi @AlexanderDokuchaev 

> Yes, i can but need some time.
Thank you, I'm waiting for your instruction.

> One question, why do you want to use grayscale image?
Yeah, I'm building a face detector for thermal camera. I trained with Yolo Tiny V3 and it worked for me. However, the bounding box isn't good enough, pretty bad and unstable one.
The bounding box from mobilenetv2_tiny_ssd300_wider_face very good. So I want to build the thermal face detection with same config as mobilenetv2_tiny_ssd300_wider_face.

Regards,
Michael

**Comment by AlexanderDokuchaev** [comment]
@michaelnguyen11 

Bad news, mmdetection does not support 1 channel images as input and convert all images to RGB.

Best way to process grayscale image as RGB. In this case you can use your model that you already have (1 channels images was read as RGB, where R = G = B).
On inference you can read image by `cv::imread(img_path)` or convert to RGB by `cv::cvtColor(gray, img, CV_GRAY2BGR)`, and put 3 channels image to input blob. 

Alternative way is really hard, need modify DataLoader to have grayscale image in the end of pipeline, change code for MobileNetV2, change pretrained weights for the first layer and possible something else. And it should not significantly improve performance or accuracy.

**Comment by michaelnguyen11** [comment]
Hi @AlexanderDokuchaev ,

Thank you very much for your effort and time.

From this ticket https://github.com/open-mmlab/mmdetection/issues/178 , look like the "mmcv.imread" automatically convert grayscale(num_channel=1) to color(num_channel=3)

I tried your intstruction : the gray image to BGR image and put 3 channels image as input blob, and it works like a charm.

Thank again for your support.

Regards,
Michael
